File: packages/protocol/src/events.ts
import { ZKQ_EVENT_SCHEMA_V1 } from "./version.js";
import type { BallotCommitment, ElectionId, NullifierHash, RootHash, TallyBucket, Sha256Hex } from "./ids.js";

export const EVENT_NAME_VOTE_CAST = "VoteCastV1" as const;
export const EVENT_NAME_VOTE_COMMITTED = "VoteCommittedV1" as const;
export const EVENT_NAME_VOTE_REVEALED = "VoteRevealedV1" as const;

export type EventName =
  | typeof EVENT_NAME_VOTE_CAST
  | typeof EVENT_NAME_VOTE_COMMITTED
  | typeof EVENT_NAME_VOTE_REVEALED;

export interface EventEnvelope<T> {
  readonly schema: typeof ZKQ_EVENT_SCHEMA_V1;
  readonly name: EventName;
  readonly electionId: ElectionId;
  readonly payload: T;
}

export interface VoteCastV1Payload {
  readonly nullifierHash: NullifierHash;
  readonly vote: number;
  readonly tallyBucket: TallyBucket;
  readonly publicSchemaVersion: string;
  readonly proofHash: Sha256Hex;
  readonly publicSignalsHash: Sha256Hex;
  readonly stateRoot: RootHash;
  readonly associationRoot: RootHash;
}

export interface VoteCommittedV1Payload {
  readonly nullifierHash: NullifierHash;
  readonly ballotCommitment: BallotCommitment;
  readonly tallyBucket: TallyBucket;
  readonly publicSchemaVersion: string;
  readonly proofHash: Sha256Hex;
  readonly publicSignalsHash: Sha256Hex;
  readonly stateRoot: RootHash;
  readonly associationRoot: RootHash;
}

export interface VoteRevealedV1Payload {
  readonly electionId: ElectionId;
  readonly ballotCommitment: BallotCommitment;
  readonly vote: number;
}

export type VoteCastV1 = EventEnvelope<VoteCastV1Payload>;
export type VoteCommittedV1 = EventEnvelope<VoteCommittedV1Payload>;
export type VoteRevealedV1 = EventEnvelope<VoteRevealedV1Payload>;

export type ZkQuorumEvent = VoteCastV1 | VoteCommittedV1 | VoteRevealedV1;

File: contracts/zk-quorum/src/events.rs
use soroban_sdk::{contractevent, BytesN, Env};

#[contractevent]
pub struct VoteCastV1 {
    pub election_id: BytesN<32>,
    pub nullifier_hash: BytesN<32>,
    pub vote: u32,
    pub tally_bucket: u32,
    pub public_schema_version: u32,
    pub proof_hash: BytesN<32>,
    pub public_signals_hash: BytesN<32>,
}

#[contractevent]
pub struct VoteCommittedV1 {
    pub election_id: BytesN<32>,
    pub nullifier_hash: BytesN<32>,
    pub ballot_commitment: BytesN<32>,
    pub tally_bucket: u32,
    pub public_schema_version: u32,
    pub proof_hash: BytesN<32>,
    pub public_signals_hash: BytesN<32>,
}

#[contractevent]
pub struct VoteRevealedV1 {
    pub election_id: BytesN<32>,
    pub ballot_commitment: BytesN<32>,
    pub vote: u32,
}

#[allow(clippy::too_many_arguments)]
pub fn emit_vote_cast_v1(
    env: &Env,
    election_id: &BytesN<32>,
    nullifier_hash: &BytesN<32>,
    vote: u32,
    tally_bucket: u32,
    public_schema_version: u32,
    proof_hash: &BytesN<32>,
    public_signals_hash: &BytesN<32>,
) {
    env.events().publish_event(&VoteCastV1 {
        election_id: election_id.clone(),
        nullifier_hash: nullifier_hash.clone(),
        vote,
        tally_bucket,
        public_schema_version,
        proof_hash: proof_hash.clone(),
        public_signals_hash: public_signals_hash.clone(),
    });
}

#[allow(clippy::too_many_arguments)]
pub fn emit_vote_committed_v1(
    env: &Env,
    election_id: &BytesN<32>,
    nullifier_hash: &BytesN<32>,
    ballot_commitment: &BytesN<32>,
    tally_bucket: u32,
    public_schema_version: u32,
    proof_hash: &BytesN<32>,
    public_signals_hash: &BytesN<32>,
) {
    env.events().publish_event(&VoteCommittedV1 {
        election_id: election_id.clone(),
        nullifier_hash: nullifier_hash.clone(),
        ballot_commitment: ballot_commitment.clone(),
        tally_bucket,
        public_schema_version,
        proof_hash: proof_hash.clone(),
        public_signals_hash: public_signals_hash.clone(),
    });
}

pub fn emit_vote_revealed_v1(
    env: &Env,
    election_id: &BytesN<32>,
    ballot_commitment: &BytesN<32>,
    vote: u32,
) {
    env.events().publish_event(&VoteRevealedV1 {
        election_id: election_id.clone(),
        ballot_commitment: ballot_commitment.clone(),
        vote,
    });
}
